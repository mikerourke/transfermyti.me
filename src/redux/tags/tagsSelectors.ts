import { isNil, propOr } from "ramda";

import { mappingByToolNameSelector } from "~/redux/allEntities/allEntitiesSelectors";
import { createSelector, type Selector } from "~/redux/reduxTools";
import { sourceTimeEntryCountByTagIdSelector } from "~/redux/timeEntries/timeEntriesSelectors";
import { activeWorkspaceIdSelector } from "~/redux/workspaces/workspacesSelectors";
import type {
  EntityTableRecord,
  Mapping,
  ReduxState,
  Tag,
  ToolName,
} from "~/types";

const sourceTagsByIdSelector = createSelector(
  (state: ReduxState) => state.tags.source,
  (sourceTagsById): Dictionary<Tag> => sourceTagsById,
);

const sourceTagsSelector = createSelector(
  sourceTagsByIdSelector,
  (sourceTagsById): Tag[] => Object.values(sourceTagsById),
);

const targetTagsByIdSelector = createSelector(
  (state: ReduxState) => state.tags.target,
  (targetTagsById): Dictionary<Tag> => targetTagsById,
);

const targetTagsSelector = createSelector(
  targetTagsByIdSelector,
  (targetTagsById): Tag[] => Object.values(targetTagsById),
);

const tagsByMappingSelector = createSelector(
  sourceTagsSelector,
  targetTagsSelector,
  (sourceTags, targetTags): Record<Mapping, Tag[]> => ({
    source: sourceTags,
    target: targetTags,
  }),
);

export const includedSourceTagsSelector = createSelector(
  sourceTagsSelector,
  (sourceTags): Tag[] => sourceTags.filter((sourceTag) => sourceTag.isIncluded),
);

export const sourceTagsForTransferSelector = createSelector(
  includedSourceTagsSelector,
  (sourceTags): Tag[] =>
    sourceTags.filter((sourceTag) => isNil(sourceTag.linkedId)),
);

const sourceTagsInActiveWorkspaceSelector = createSelector(
  sourceTagsSelector,
  activeWorkspaceIdSelector,
  (sourceTags, workspaceId): Tag[] =>
    sourceTags.filter((tag) => tag.workspaceId === workspaceId),
);

export const tagsForInclusionsTableSelector = createSelector(
  (state: ReduxState) => state.allEntities.areExistsInTargetShown,
  sourceTagsInActiveWorkspaceSelector,
  sourceTimeEntryCountByTagIdSelector,
  (
    areExistsInTargetShown,
    sourceTags,
    timeEntryCountByTagId,
  ): EntityTableRecord<Tag>[] => {
    const tagTableRecords: EntityTableRecord<Tag>[] = [];

    for (const sourceTag of sourceTags) {
      const existsInTarget = sourceTag.linkedId !== null;
      if (existsInTarget && !areExistsInTargetShown) {
        continue;
      }

      const entryCount = propOr<number, Dictionary<number>, number>(
        0,
        sourceTag.id,
        timeEntryCountByTagId,
      );

      tagTableRecords.push({
        ...sourceTag,
        entryCount,
        existsInTarget,
        isActiveInSource: true,
        isActiveInTarget: existsInTarget,
      });
    }

    return tagTableRecords;
  },
);

export const tagsTotalCountsByTypeSelector = createSelector(
  tagsForInclusionsTableSelector,
  (tagsForInclusionTable) =>
    tagsForInclusionTable.reduce(
      (
        acc,
        { entryCount, existsInTarget, isIncluded }: EntityTableRecord<Tag>,
      ) => ({
        entryCount: acc.entryCount + entryCount,
        existsInTarget: acc.existsInTarget + (existsInTarget ? 1 : 0),
        isIncluded: acc.isIncluded + (isIncluded ? 1 : 0),
      }),
      {
        entryCount: 0,
        existsInTarget: 0,
        isIncluded: 0,
      },
    ),
);

export const tagIdsByNameBySelectorFactory = (
  toolName: ToolName,
): Selector<ReduxState, Dictionary<string>> =>
  createSelector(
    mappingByToolNameSelector,
    tagsByMappingSelector,
    (mappingByToolName, tagsByMapping): Dictionary<string> => {
      const toolMapping = mappingByToolName[toolName];

      const tags = tagsByMapping[toolMapping];

      const tagIdsByName: Dictionary<string> = {};
      for (const tag of tags) {
        tagIdsByName[tag.name] = tag.id;
      }

      return tagIdsByName;
    },
  );

export const targetTagIdsSelectorFactory = (
  sourceTagIds: string[],
): Selector<ReduxState, string[]> =>
  createSelector(sourceTagsByIdSelector, (sourceTagsById): string[] => {
    const targetTagIds: string[] = [];

    for (const sourceTagId of sourceTagIds) {
      const sourceTag = sourceTagsById[sourceTagId];

      if (sourceTag && sourceTag.linkedId) {
        targetTagIds.push(sourceTag.linkedId);
      }
    }

    return targetTagIds;
  });
