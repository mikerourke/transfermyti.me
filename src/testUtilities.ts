export { FAKES } from "~/redux/__fakes__/state";

type MockInstance = jest.MockInstance<AnyValid, AnyValid>;

interface WrapForMockingReturn {
  calls: AnyValid[];
  clear(): MockInstance;
  implementOnce(funcOrReturn: AnyValid): MockInstance;
  rejectAlways(err?: AnyValid): MockInstance;
  rejectOnce(err?: AnyValid): MockInstance;
  resolveMultiple(values: AnyValid[]): void;
  resolveOnce(value?: AnyValid): MockInstance;
  returnOnce(value?: AnyValid): MockInstance;
  returnMultiple(values: AnyValid[]): void;
  throwOnce(message?: string): MockInstance;
}

/**
 * Wrappers around someFunc.mock...() that Jest provides. Using
 * this avoids the annoying `someFunc as AnyValid` required to shut TypeScript
 * up.
 */
export function wrapForMocking(func: AnyValid): WrapForMockingReturn {
  return {
    calls: func?.mock?.calls,
    clear(): MockInstance {
      return func.mockClear();
    },
    implementOnce(funcOrReturn: AnyValid): MockInstance {
      if (typeof funcOrReturn === "function") {
        return func.mockImplementationOnce(funcOrReturn);
      } else {
        return func.mockImplementationOnce(() => funcOrReturn);
      }
    },
    rejectAlways(err: AnyValid = "error"): MockInstance {
      return func.mockRejectedValue(
        typeof err === "string" ? new Error(err) : err,
      );
    },
    rejectOnce(err: AnyValid = "error"): MockInstance {
      return func.mockRejectedValueOnce(
        typeof err === "string" ? new Error(err) : err,
      );
    },
    resolveMultiple(values: AnyValid[]): void {
      for (const value of values) {
        func.mockResolvedValueOnce(value);
      }
    },
    resolveOnce(value: AnyValid = null): MockInstance {
      return func.mockResolvedValueOnce(value);
    },
    returnOnce(value: AnyValid = null): MockInstance {
      return func.mockReturnValueOnce(value);
    },
    returnMultiple(values: AnyValid[]): void {
      for (const value of values) {
        func.mockReturnValueOnce(value);
      }
    },
    throwOnce(message: string = "error"): MockInstance {
      return func.mockImplementationOnce(() => {
        throw new Error(message);
      });
    },
  };
}

/**
 * Allows you to get around `delay()` calls in sagas.
 * @see https://github.com/jfairbank/redux-saga-test-plan/issues/257#issuecomment-503016758
 * @example
 *   return expectSaga(someSaga)
 *     .provide([
 *       { call: provideDelay },
 *     ])
 *     ...
 */
export function provideDelay(
  { fn }: AnyValid,
  next: AnyValid,
): null | VoidFunction {
  return fn.name === "delayP" ? null : next();
}
