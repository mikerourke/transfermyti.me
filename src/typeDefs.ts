import type { AllEntitiesState } from "~/redux/allEntities/allEntitiesReducer";
import type { AppState } from "~/redux/app/appReducer";
import type { ClientsState } from "~/redux/clients/clientsReducer";
import type { CredentialsState } from "~/redux/credentials/credentialsReducer";
import type { ProjectsState } from "~/redux/projects/projectsReducer";
import type { TagsState } from "~/redux/tags/tagsReducer";
import type { TasksState } from "~/redux/tasks/tasksReducer";
import type { TimeEntriesState } from "~/redux/timeEntries/timeEntriesReducer";
import type { UserGroupsState } from "~/redux/userGroups/userGroupsReducer";
import type { UsersState } from "~/redux/users/usersReducer";
import type { WorkspacesState } from "~/redux/workspaces/workspacesReducer";

export interface ReduxState {
  allEntities: AllEntitiesState;
  app: AppState;
  clients: ClientsState;
  credentials: CredentialsState;
  projects: ProjectsState;
  tags: TagsState;
  tasks: TasksState;
  timeEntries: TimeEntriesState;
  userGroups: UserGroupsState;
  users: UsersState;
  workspaces: WorkspacesState;
}

type NotificationType = "error" | "info" | "success";

export interface Notification {
  id: string;
  message: string;
  type: NotificationType;
}

export enum ToolName {
  None = "none",
  Clockify = "clockify",
  Toggl = "toggl",
}

export enum Mapping {
  Source = "source",
  Target = "target",
}

export enum ToolAction {
  None = "none",
  Delete = "delete",
  Transfer = "transfer",
}

export interface ToolHelpDetails {
  toolName: ToolName;
  displayName: string;
  toolLink: string;
}

export enum EntityGroup {
  Clients = "clients",
  Projects = "projects",
  Tags = "tags",
  Tasks = "tasks",
  TimeEntries = "timeEntries",
  UserGroups = "userGroups",
  Users = "users",
  Workspaces = "workspaces",
}

export enum FetchStatus {
  Pending = "PENDING",
  InProcess = "IN_PROCESS",
  Success = "SUCCESS",
  Error = "ERROR",
}

export interface AnyEntity {
  id: string;
  workspaceId: string;
  linkedId: string | null;
  isIncluded: boolean;
  entryCount: number;
  memberOf?: EntityGroup;
}

export type ValidEntity<TEntity> = TEntity & AnyEntity;

export type EntityTableRecord<TEntity> = TEntity & {
  existsInTarget: boolean;
  isActiveInSource: boolean;
  isActiveInTarget: boolean;
};

export type CountsByEntityGroup = Record<
  Exclude<EntityGroup, EntityGroup.Workspaces>,
  number
>;

export interface ToolNameByMapping {
  [Mapping.Source]: ToolName;
  [Mapping.Target]: ToolName;
}

export interface Client extends AnyEntity {
  id: string;
  name: string;
  workspaceId: string;
}

export type ClientTableRecord = EntityTableRecord<Client> & {
  projectCount: number;
};

export interface Credentials {
  apiKey: string | null;
  email: string | null;
  userId: string | null;
}

export type CredentialsByMapping = Record<Mapping, Credentials>;

export type ValidationErrorsByMapping = Record<Mapping, string | null>;

export interface PartialCredentialsUpdate {
  mapping: Mapping;
  apiKey?: string | null;
  email?: string | null;
  userId?: string | null;
}

export interface Estimate {
  estimate: number | string;
  type: "AUTO" | "MANUAL";
}

export interface Project extends AnyEntity {
  id: string;
  name: string;
  clientId: string | null;
  isBillable: boolean;
  isPublic: boolean;
  isActive: boolean;
  estimate: Estimate;
  color: string;
  userIds: string[];
}

export interface Tag extends AnyEntity {
  id: string;
  name: string;
  workspaceId: string;
}

export interface Task extends AnyEntity {
  id: string;
  name: string;
  estimate: string;
  projectId: string;
  assigneeIds: string[];
  isActive: boolean;
}

export type TaskTableRecord = EntityTableRecord<Task> & {
  projectName: string;
};

export interface TimeEntry extends AnyEntity {
  id: string;
  description: string;
  isBillable: boolean;
  start: Date;
  end: Date;
  year: number;
  isActive: boolean;
  clientId: string | null;
  projectId: string | null;
  tagIds: string[];
  tagNames: string[];
  taskId: string | null;
  userId: string | null;
  userGroupIds: string[];
}

export interface TimeEntryTableRecord extends EntityTableRecord<TimeEntry> {
  taskName: string | null;
  projectName: string | null;
}

export interface UserGroup extends AnyEntity {
  id: string;
  name: string;
  userIds: string[];
}

export interface User extends AnyEntity {
  id: string;
  name: string;
  email: string;
  isAdmin: boolean | null;
  isActive: boolean;
  userGroupIds: string[] | null;
}

export interface Workspace extends AnyEntity {
  id: string;
  name: string;
  userIds: string[];
  isAdmin: boolean | null;
}
