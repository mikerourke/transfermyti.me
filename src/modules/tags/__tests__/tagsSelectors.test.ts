import cases from "jest-in-case";

import { FAKES } from "~/testUtilities";
import { ToolName } from "~/typeDefs";

import * as tagsSelectors from "../tagsSelectors";

const MOCK_STATE = {
  ...FAKES.REDUX_STATE,
  tags: {
    ...FAKES.REDUX_STATE.tags,
    source: {
      ...FAKES.REDUX_STATE.tags.source,
      "4004": {
        ...FAKES.REDUX_STATE.tags.source["4004"],
        linkedId: null,
        isIncluded: false,
      },
      "4005": {
        ...FAKES.REDUX_STATE.tags.source["4005"],
        linkedId: null,
        isIncluded: false,
      },
      "4006": {
        ...FAKES.REDUX_STATE.tags.source["4006"],
        linkedId: null,
        isIncluded: true,
      },
      "4007": {
        ...FAKES.REDUX_STATE.tags.source["4007"],
        linkedId: null,
        isIncluded: true,
      },
      "4008": {
        ...FAKES.REDUX_STATE.tags.source["4008"],
        linkedId: null,
        isIncluded: true,
      },
      "4009": {
        ...FAKES.REDUX_STATE.tags.source["4009"],
        linkedId: null,
        isIncluded: true,
      },
    },
    target: {
      ...FAKES.REDUX_STATE.tags.target,
    },
  },
};

describe("within tagsSelectors", () => {
  cases(
    "the selectors match their snapshots",
    (options) => {
      const result = options.selector(MOCK_STATE);

      expect(result).toMatchSnapshot();
    },
    [
      {
        name: "for the includedSourceTagsSelector",
        selector: tagsSelectors.includedSourceTagsSelector,
      },
      {
        name: "for the sourceTagsForTransferSelector",
        selector: tagsSelectors.sourceTagsForTransferSelector,
      },
      {
        name: "for the tagsTotalCountsByTypeSelector",
        selector: tagsSelectors.tagsTotalCountsByTypeSelector,
      },
    ],
  );

  cases(
    "the tagsForInclusionsTableSelector matches its snapshot based on state.allEntities.areExistsInTargetShown",
    (options) => {
      const updatedState = {
        ...MOCK_STATE,
        allEntities: {
          ...MOCK_STATE.allEntities,
          areExistsInTargetShown: options.areExistsInTargetShown,
        },
      };

      const result = tagsSelectors.tagsForInclusionsTableSelector(updatedState);

      expect(result).toMatchSnapshot();
    },
    [
      {
        name: "when state.allEntities.areExistsInTargetShown = true",
        areExistsInTargetShown: true,
      },
      {
        name: "when state.allEntities.areExistsInTargetShown = false",
        areExistsInTargetShown: false,
      },
    ],
  );

  test("the tagIdsByNameSelectorFactory matches its snapshot", () => {
    const getTagIdsByName = tagsSelectors.tagIdsByNameBySelectorFactory(ToolName.Toggl);

    const result = getTagIdsByName(MOCK_STATE);

    expect(result).toMatchSnapshot();
  });

  test("the targetTagIdsSelectorFactory matches its snapshot", () => {
    const getTargetTagIds = tagsSelectors.targetTagIdsSelectorFactory([
      "4001",
      "4002",
      "4003",
      "4004",
      "4005",
    ]);

    const result = getTargetTagIds(MOCK_STATE);

    expect(result).toMatchSnapshot();
  });
});
