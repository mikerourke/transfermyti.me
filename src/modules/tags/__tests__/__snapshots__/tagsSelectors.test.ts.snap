// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`within tagsSelectors the selectors match their snapshots for the includedSourceTagsSelector 1`] = `
[
  {
    "entryCount": 0,
    "id": "4006",
    "isIncluded": true,
    "linkedId": null,
    "memberOf": "tags",
    "name": "tag-f",
    "workspaceId": "1001",
  },
  {
    "entryCount": 0,
    "id": "4007",
    "isIncluded": true,
    "linkedId": null,
    "memberOf": "tags",
    "name": "tag-g",
    "workspaceId": "1001",
  },
  {
    "entryCount": 0,
    "id": "4008",
    "isIncluded": true,
    "linkedId": null,
    "memberOf": "tags",
    "name": "tag-h",
    "workspaceId": "1001",
  },
  {
    "entryCount": 0,
    "id": "4009",
    "isIncluded": true,
    "linkedId": null,
    "memberOf": "tags",
    "name": "tag-i",
    "workspaceId": "1001",
  },
]
`;

exports[`within tagsSelectors the selectors match their snapshots for the sourceTagsForTransferSelector 1`] = `
[
  {
    "entryCount": 0,
    "id": "4006",
    "isIncluded": true,
    "linkedId": null,
    "memberOf": "tags",
    "name": "tag-f",
    "workspaceId": "1001",
  },
  {
    "entryCount": 0,
    "id": "4007",
    "isIncluded": true,
    "linkedId": null,
    "memberOf": "tags",
    "name": "tag-g",
    "workspaceId": "1001",
  },
  {
    "entryCount": 0,
    "id": "4008",
    "isIncluded": true,
    "linkedId": null,
    "memberOf": "tags",
    "name": "tag-h",
    "workspaceId": "1001",
  },
  {
    "entryCount": 0,
    "id": "4009",
    "isIncluded": true,
    "linkedId": null,
    "memberOf": "tags",
    "name": "tag-i",
    "workspaceId": "1001",
  },
]
`;

exports[`within tagsSelectors the selectors match their snapshots for the tagsTotalCountsByTypeSelector 1`] = `
{
  "entryCount": 3,
  "existsInTarget": 3,
  "isIncluded": 4,
}
`;

exports[`within tagsSelectors the tagIdsByNameSelectorFactory matches its snapshot 1`] = `
{
  "tag-a": "4001",
  "tag-b": "4002",
  "tag-c": "4003",
  "tag-d": "4004",
  "tag-e": "4005",
  "tag-f": "4006",
  "tag-g": "4007",
  "tag-h": "4008",
  "tag-i": "4009",
}
`;

exports[`within tagsSelectors the tagsForInclusionsTableSelector matches its snapshot based on state.allEntities.areExistsInTargetShown when state.allEntities.areExistsInTargetShown = false 1`] = `
[
  {
    "entryCount": 0,
    "existsInTarget": false,
    "id": "4004",
    "isActiveInSource": true,
    "isActiveInTarget": false,
    "isIncluded": false,
    "linkedId": null,
    "memberOf": "tags",
    "name": "tag-d",
    "workspaceId": "1001",
  },
  {
    "entryCount": 0,
    "existsInTarget": false,
    "id": "4005",
    "isActiveInSource": true,
    "isActiveInTarget": false,
    "isIncluded": false,
    "linkedId": null,
    "memberOf": "tags",
    "name": "tag-e",
    "workspaceId": "1001",
  },
  {
    "entryCount": 0,
    "existsInTarget": false,
    "id": "4006",
    "isActiveInSource": true,
    "isActiveInTarget": false,
    "isIncluded": true,
    "linkedId": null,
    "memberOf": "tags",
    "name": "tag-f",
    "workspaceId": "1001",
  },
  {
    "entryCount": 0,
    "existsInTarget": false,
    "id": "4007",
    "isActiveInSource": true,
    "isActiveInTarget": false,
    "isIncluded": true,
    "linkedId": null,
    "memberOf": "tags",
    "name": "tag-g",
    "workspaceId": "1001",
  },
  {
    "entryCount": 0,
    "existsInTarget": false,
    "id": "4008",
    "isActiveInSource": true,
    "isActiveInTarget": false,
    "isIncluded": true,
    "linkedId": null,
    "memberOf": "tags",
    "name": "tag-h",
    "workspaceId": "1001",
  },
  {
    "entryCount": 0,
    "existsInTarget": false,
    "id": "4009",
    "isActiveInSource": true,
    "isActiveInTarget": false,
    "isIncluded": true,
    "linkedId": null,
    "memberOf": "tags",
    "name": "tag-i",
    "workspaceId": "1001",
  },
]
`;

exports[`within tagsSelectors the tagsForInclusionsTableSelector matches its snapshot based on state.allEntities.areExistsInTargetShown when state.allEntities.areExistsInTargetShown = true 1`] = `
[
  {
    "entryCount": 1,
    "existsInTarget": true,
    "id": "4001",
    "isActiveInSource": true,
    "isActiveInTarget": true,
    "isIncluded": false,
    "linkedId": "clock-tag-01",
    "memberOf": "tags",
    "name": "tag-a",
    "workspaceId": "1001",
  },
  {
    "entryCount": 1,
    "existsInTarget": true,
    "id": "4002",
    "isActiveInSource": true,
    "isActiveInTarget": true,
    "isIncluded": false,
    "linkedId": "clock-tag-02",
    "memberOf": "tags",
    "name": "tag-b",
    "workspaceId": "1001",
  },
  {
    "entryCount": 1,
    "existsInTarget": true,
    "id": "4003",
    "isActiveInSource": true,
    "isActiveInTarget": true,
    "isIncluded": false,
    "linkedId": "clock-tag-03",
    "memberOf": "tags",
    "name": "tag-c",
    "workspaceId": "1001",
  },
  {
    "entryCount": 0,
    "existsInTarget": false,
    "id": "4004",
    "isActiveInSource": true,
    "isActiveInTarget": false,
    "isIncluded": false,
    "linkedId": null,
    "memberOf": "tags",
    "name": "tag-d",
    "workspaceId": "1001",
  },
  {
    "entryCount": 0,
    "existsInTarget": false,
    "id": "4005",
    "isActiveInSource": true,
    "isActiveInTarget": false,
    "isIncluded": false,
    "linkedId": null,
    "memberOf": "tags",
    "name": "tag-e",
    "workspaceId": "1001",
  },
  {
    "entryCount": 0,
    "existsInTarget": false,
    "id": "4006",
    "isActiveInSource": true,
    "isActiveInTarget": false,
    "isIncluded": true,
    "linkedId": null,
    "memberOf": "tags",
    "name": "tag-f",
    "workspaceId": "1001",
  },
  {
    "entryCount": 0,
    "existsInTarget": false,
    "id": "4007",
    "isActiveInSource": true,
    "isActiveInTarget": false,
    "isIncluded": true,
    "linkedId": null,
    "memberOf": "tags",
    "name": "tag-g",
    "workspaceId": "1001",
  },
  {
    "entryCount": 0,
    "existsInTarget": false,
    "id": "4008",
    "isActiveInSource": true,
    "isActiveInTarget": false,
    "isIncluded": true,
    "linkedId": null,
    "memberOf": "tags",
    "name": "tag-h",
    "workspaceId": "1001",
  },
  {
    "entryCount": 0,
    "existsInTarget": false,
    "id": "4009",
    "isActiveInSource": true,
    "isActiveInTarget": false,
    "isIncluded": true,
    "linkedId": null,
    "memberOf": "tags",
    "name": "tag-i",
    "workspaceId": "1001",
  },
]
`;

exports[`within tagsSelectors the targetTagIdsSelectorFactory matches its snapshot 1`] = `
[
  "clock-tag-01",
  "clock-tag-02",
  "clock-tag-03",
]
`;
